tycho-indexer/src/extractor/dynamic_contract_indexer/dci.rs

```diff
@@ -553,21 +571,32 @@ where
             self.update_cache(&block_changes.block, &traced_entry_points)?;
             drop(_span);
 
-            // Store failed traces as None in the cache
+            // Store failed traces as None in the cache and increment retry counter
             for (failed_ep, _) in failed_entrypoints.iter() {
+                let key = (
+                    failed_ep
+                        .entry_point
+                        .external_id
+                        .clone(),
+                    failed_ep.params.clone(),
+                );
+
+                // Store the failed trace result
                 self.cache
                     .entrypoint_results
-                    .insert_pending(
-                        block_changes.block.clone(),
-                        (
-                            failed_ep
-                                .entry_point
-                                .external_id
-                                .clone(),
-                            failed_ep.params.clone(),
-                        ),
-                        None,
-                    )?;
+                    .insert_pending(block_changes.block.clone(), key.clone(), None)?;
+
+                // Get current retry count (default to 0 if first failure) and increment
+                let current_retry_count = self
+                    .cache
+                    .tracing_retry_counts
+                    .get(&key)
+                    .cloned()
+                    .unwrap_or(0);
+
+                self.cache
+                    .tracing_retry_counts
+                    .insert_pending(block_changes.block.clone(), key, current_retry_count + 1)?;
```

Comment:
Would be nice to combine this so that you only access the cache once and remove the need for temp vars. Maybe something like this (just a suggestion, it likely needs slight adjusting):
```suggestion
                let current_retry_count = self
                    .cache
                    .tracing_retry_counts
                    .pending_entry(&block_changes.block, &key)?
                    .and_modify(|count| *count += 1)
                    .or_insert(1);
```